pkg/tsdb/prometheus/time_series_query.go

```diff
@@ -299,15 +299,30 @@ func matrixToDataFrames(matrix model.Matrix, query *PrometheusQuery, frames data
 			tags[string(k)] = string(v)
 		}
 
-		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, len(v.Values))
-		valueField := data.NewFieldFromFieldType(data.FieldTypeNullableFloat64, len(v.Values))
+		baseTimestamp := alignTimeRange(query.Start, query.Step, query.UtcOffsetSec).UnixMilli()
+		endTimestamp := alignTimeRange(query.End, query.Step, query.UtcOffsetSec).UnixMilli()
+		// For each step we create 1 data point. This results in range / step + 1 data points.
+		datapointsCount := int((endTimestamp-baseTimestamp)/query.Step.Milliseconds()) + 1
 
-		for i, k := range v.Values {
-			timeField.Set(i, time.Unix(k.Timestamp.Unix(), 0).UTC())
-			value := float64(k.Value)
+		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, datapointsCount)
+		valueField := data.NewFieldFromFieldType(data.FieldTypeNullableFloat64, datapointsCount)
+		idx := 0
+
+		for _, pair := range v.Values {
+			timestamp := int64(pair.Timestamp)
+			value := float64(pair.Value)
+
+			for t := baseTimestamp; t < timestamp; t += query.Step.Milliseconds() {
+				timeField.Set(idx, time.Unix(0, t*1000000).UTC())
+				idx++
+			}
+
+			timeField.Set(idx, time.Unix(pair.Timestamp.Unix(), 0).UTC())
```

Comment:
i'm not sure of the best way to phrase this question, so let me know if this doesn't make sense: is it safe to assume that `pair.Timestamp.Unix()` will match the step spacing of the loop above?

Reply:
Yeah, good question! I would say it is safe as we were [doing the same thing on frontend](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/result_transformer.ts#L383-L389). Moreover, that's how Prometheus returns data - for each step 1 data point and they are exactly 1 step spaced.